File: javascript/src/leetcode/_I_0409_H.ts

class TreeNode {
    val: number;
    left: TreeNode;
    right: TreeNode;
    constructor(val: number) {
        this.val = val;
    }
}

function visit(node: TreeNode, queue: TreeNode[], path: number[], res: number[][]) {
    path.push(node.val);
    if (node.left != null) {
        queue.push(node.left);
    }
    if (node.right != null) {
        queue.push(node.right);
    }
    if (queue.length == 0) {
        res.push(Array.from(path));
    }
    else {
        for (let i = 0, len = queue.length; i < len; i++) {
            let cur = queue.splice(i, 1)[0];
            visit(cur, queue, path, res);
            queue.splice(i, 0, cur);
        }
        if (node.right != null) {
            queue.pop();
        }
        if (node.left != null) {
            queue.pop();
        }
    }
    path.pop();
}

function BSTSequences(root: TreeNode) {
    if (root == null) {
        return [[]];
    }
    const res = [];
    visit(root, [], [], res);
    return res;
}

const root = new TreeNode(5);
root.left = new TreeNode(2);
root.left.left = new TreeNode(1);
root.left.right = new TreeNode(4);
root.left.left.left = new TreeNode(3);
const lists = BSTSequences(root);
console.log(lists);

File: python/leetcode/_I_0409_H.py
from typing import List


class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None


class Solution:

    def visit(self, node: TreeNode, queue: List[TreeNode], path: List[int], res: List[List[int]]):
        path.append(node.val)
        if node.left:
            queue.append(node.left)
        if node.right:
            queue.append(node.right)
        if len(queue) == 0:
            res.append(path.copy())
        else:
            for i, cur in enumerate(queue):
                self.visit(cur, queue[:i] + queue[i + 1:], path, res)
            if node.right:
                queue.pop()
            if node.left:
                queue.pop()
        path.pop()

    def BSTSequences(self, root: TreeNode) -> List[List[int]]:
        if root is None:
            return [[]]
        res = []
        self.visit(root, [], [], res)
        return res


if __name__ == '__main__':
    root = TreeNode(5)
    root.left = TreeNode(2)
    root.left.left = TreeNode(1)
    root.left.right = TreeNode(4)
    root.left.left.left = TreeNode(3)
    lists = Solution().BSTSequences(root)
    print(lists)
